Miovision Aggregation Improvements (ATR view, intersection_movements AFTER INSERT trigger, unacceptable_gaps)

gabrielwol opened the pull request:
## What this pull request accomplishes:

1. Remove `miovision_api.volumes_15min` (table) in place of view (`volumes_15min_atr_filtered`). Reduces duplicate storage (~75GB) and complexity in aggregation. 
2. Add an `AFTER INSERT` trigger to `intersection_movements` to help aggregate and pad new movements into volumes_15min_mvt.  
    - volumes/miovision/sql/function/function-add_intersection_movement_padding_values.sql: based on existing inserts into volumes_15min_mvt, except using `new_rows` (transition relation) instead of `intersection_movements` and without date limits. 
3. Move `unacceptable_gap` removal from table inserts to filtered view: less complexity in adding and removing unacceptable_gaps. 

**Summary of files changed:**
Many files have minor changes including renamed references, removal of ATR aggregation, big fluff (intersection_movements, volumes_15min_tmc).

Files with major changes:
- [volumes/miovision/sql/function/function-add_intersection_movement_padding_values.sql](https://github.com/CityofToronto/bdit_data-sources/pull/1024/files#diff-a18eb22409f0b923d050598f95788f37425c53d02664af98a4e20d2f0b85ad57): trigger function to add values to 15min aggregate table based on additions to intersection_movements. Based closely on [existing aggregation](https://github.com/CityofToronto/bdit_data-sources/blob/master/volumes/miovision/sql/function/function-aggregate-volumes_15min_mvt.sql) but only for new intersection_movements. 
- [volumes/miovision/sql/table/create-table-movement_map_new.sql](https://github.com/CityofToronto/bdit_data-sources/pull/1024/files#diff-436002ac1093f53236f22451598bb06ff4c7eb1edf1b622c49677f6fa6f5cd48): a wide version of `movement_map` with entry/exit leg columns and directions more clearly labelled. Helps with ATR view joins. 
- [volumes/miovision/sql/views/create-view-volumes_15min_atr_filtered.sql](https://github.com/CityofToronto/bdit_data-sources/pull/1024/files#diff-a25a0ab27d992960339211265d569955314948c953008335fcd37c969f0f74de): ATR view which replaces table. 
- [volumes/miovision/sql/views/create-view-volumes_15min_mvt_filtered.sql](https://github.com/CityofToronto/bdit_data-sources/pull/1024/files#diff-1eff77b42680121ec7dc4802db27ff6ec8eac64ea82130cf772ac90c99f5a2d4): add anti-join of unacceptable_gaps. 

## Issue(s) this solves:

- Closes #868, #879: filters `unacceptable_gaps` via view
- Progress towards closing issues around `intersection_movements`: #406, #855

## What, in particular, needs to reviewed:

- [ ] Another option considered is adding movement map columns `entry_dir, exit_leg, exit_dir` to TMC table so [ATR view](https://github.com/CityofToronto/bdit_data-sources/pull/1024/files#diff-64d633fb22ceb99236e2499241f863008aeffc3605a02fd84e4ca99ec963e000) wouldn't need to join to movement_map. Downside: more confusing TMC table, does not match current TMC table. 
- [x] Ok to anti join instead of set unacceptable_gaps to null in filtered view? 
- [x] New naming: match existing to reduce dependency changes or make contents more explicit?

|     | Old                        | New                          |
|-----|----------------------------|------------------------------|
| TMC | volumes_15min_mvt          | volumes_15min_mvt_**unfiltered** |
|     | volumes_15min_mvt_filtered | volumes_15min_mvt_filtered   |
| ATR | volumes_15min              |                              |
|     | volumes_15min_filtered     | volumes_15min_**atr**_filtered   |

## What needs to be done by a sysadmin after this PR is merged
- [x] Implement all the sql changes (remove tables, remove functions, update functions, add views)
- [x] Update DAG

gabrielwol commented:
Testing the new ATR view for speed: 
```sql
--test: 0.2 s with primary key
SELECT *
FROM miovision_api.volumes_15min_atr_filtered
WHERE
    intersection_uid = 6
    AND classification_uid = 1
    AND datetime_bin = '2024-06-25 12:00:00'
    AND leg <> LEFT(dir, 1);

--DR i0627 test using new view
--41s for original, 1:06 for view (1.3M rows)
SELECT
    volumes.intersection_uid,
    date_trunc('hour', volumes.datetime_bin) AS datetime_bin,
    volumes.leg,
    volumes.dir,
    classifications.classification,
    SUM(volumes.volume) AS volume
FROM miovision_api.volumes_15min_atr_filtered AS volumes
INNER JOIN miovision_api.classifications USING (classification_uid)
WHERE
    volumes.classification_uid NOT IN (2, 7)
    AND volumes.datetime_bin >= '2024-01-01'
    AND volumes.datetime_bin < '2024-05-22'
    AND volumes.intersection_uid IN (76, 88, 81, 85, 84, 87, 75, 83, 86, 49, 80, 36)
GROUP BY
    volumes.intersection_uid,
    classifications.classification,
    date_trunc('hour', volumes.datetime_bin),
    volumes.leg,
    volumes.dir;
```

radumas commented:
> Ok to anti join instead of set unacceptable_gaps to null in filtered view?

We tested some aggregate functions with NULLs and they silently succeed while not considering the NULL values. There could be some benefit to graphing or if someone is looking at the VIEW output without aggregating to see the ranges of dates that are unacceptable. OTOH, I think analysts should be looking at the AR table as a starting point to their analysis.

radumas commented on volumes/miovision/sql/readme.md:
- [ ] The ERD at the top needs to be updated or deleted. Might have to ask @scann0n where the file is that created it. Or recreate in Mermaid
https://github.com/CityofToronto/bdit_data-sources/blob/master/volumes/miovision/sql/readme.md#miovision-data-relationships-at-a-glance
- [ ] The data dictionary for movement_map needs to be updated https://github.com/CityofToronto/bdit_data-sources/blob/master/volumes/miovision/sql/readme.md#movement_map
- [ ] Delete the row pertaining to `aggregate_15_min()` in the https://github.com/CityofToronto/bdit_data-sources/blob/master/volumes/miovision/sql/readme.md#aggregation-functions and **Clear Functions**
- [ ] Links are broken to the aggregate tables
       https://github.com/CityofToronto/bdit_data-sources/pull/1024/files#diff-2a53b055e4f243c0233aa9982681d3e4bdf35999ff11341b75128935af16ad84R62-R64

radumas reviewed (changes requested):
Some additional requested changes to make sure the documentation is in sync with the state of the code:

- [ ] https://github.com/CityofToronto/bdit_data-sources/blob/master/volumes/miovision/getting_started.md#where-should-i-look-for-data needs to be updated
- [ ] https://github.com/CityofToronto/bdit_data-sources/tree/master/volumes/miovision#where-can-i-access-the-data should include a link to Getting Started as a better starting location for accessing the data
- [ ] The changes requested in `miovision/sql/readme.md`

gabrielwol commented on volumes/miovision/sql/function/function-add_intersection_movement_padding_values.sql:
```diff
@@ -0,0 +1,130 @@
+        ) - interval '15 minutes', 
+        interval '15 minutes'
+    ) AS dt(datetime_bin)
+    WHERE
+        --0 padding for certain modes (padding)
+        im.classification_uid IN (1,2,6,10)
```
New issue: #1152

gabrielwol commented:
@radumas Ready for re-review. 

Since last week:
- Requested readme changes
- removed `report_dates` (we never used this, + it was an aggregation on top of removed volumes_15min table) + related functions and DAG task. 
- remove section used for testing
- replaced `movement_map` with `movement_map_new`. Switched from a query of old table to an insert. 
- Removed `volumes_mvt_atr_xover` (this table has an estimated 945M rows 😵, no longer necessary)

radumas commented on volumes/miovision/sql/table/create-table-movement_map_new.sql:
Could this table replace directly `movement_map`? Adding `_new` as a suffix is inviting trouble by needing to update an object again.

gabrielwol replied:
@radumas I renamed the table, and removed bike approach & ped exits from this table. Before they were implicitly being excluded as duplicates in a UNION step, now they are explicitly excluded from ATR aggregation.

radumas commented:
> @radumas Ready for re-review.
> 
> Since last week:
> 
>     * Requested readme changes
> 
>     * removed `report_dates` (we never used this, + it was an aggregation on top of removed volumes_15min table) + related functions and DAG task.
> 
>     * remove section used for testing
> 
>     * replaced `movement_map` with `movement_map_new`. Switched from a query of old table to an insert.
> 
>     * Removed `volumes_mvt_atr_xover` (this table has an estimated 945M rows 😵, no longer necessary)
> 
>     * Smol fluff: 'plpgsql' to plpgsql x 6

All these changes are great. My only remaining quibble is with the `_new` suffix.

radumas reviewed (approved):
Heck yeah! Looks great. Don't forget to fluff the NULLs

gabrielwol commented:
SQL Changes to make:
- [ ] volumes/miovision/sql/function/function-add_intersection_movement_padding_values.sql
- [ ] volumes/miovision/sql/function/function-aggregate-volumes_15min.sql
- [ ] volumes/miovision/sql/function/function-aggregate-volumes_15min_mvt.sql
- [ ] volumes/miovision/sql/function/function-clear-report_dates.sql
- [ ] volumes/miovision/sql/function/function-clear-volumes_15min.sql
- [ ] volumes/miovision/sql/function/function-clear-volumes_15min_mvt.sql
- [ ] volumes/miovision/sql/function/function-create_mm_nested_volumes_partitions.sql
- [ ] volumes/miovision/sql/function/function-create_yyyy_volumes_15min_partition.sql
- [ ] volumes/miovision/sql/function/function-get_report_dates.sql
- [ ] volumes/miovision/sql/function/function-identify-zero-counts.sql
- [ ] volumes/miovision/sql/table/create-table-intersection_movements.sql
- [ ] volumes/miovision/sql/table/create-table-movement_map.sql
- [ ] volumes/miovision/sql/table/create-table-movement_map_new.sql
- [ ] volumes/miovision/sql/table/create-table-report_dates.sql
- [ ] volumes/miovision/sql/table/create-table-volumes_15min.sql
- [ ] .../table/create-table-volumes_15min_mvt.sql → ...te-table-volumes_15min_mvt_unfiltered.sql
- [ ] volumes/miovision/sql/table/create-table-volumes_tmc_atr_xover.sql
- [ ] volumes/miovision/sql/updates/update-mio_volumes15min_partitioning.sql
- [ ] volumes/miovision/sql/views/create-view-volumes_15min_atr_filtered.sql
- [ ] volumes/miovision/sql/views/create-view-volumes_15min_filtered.sql
- [ ] volumes/miovision/sql/views/create-view-volumes_15min_mvt_filtered.sql
- [ ] volumes/miovision/sql/views/create-view-volumes_15min_tmc.sql

gabrielwol commented:
New rows added to intersection_movements now automatically update volumes_15min_mvt_unfiltered using a trigger 🎉 

```
NOTICE:  Wed Mar 19 15:52:27.551079 2025 UTC Done adding to 15min MVT bin based on intersection_movement new rows. 1924 rows added.
INSERT 0 1

Query returned successfully in 5 secs 862 msec.
```

gabrielwol commented:
to do: 
- [ ] insert data during `unacceptable_gaps` into `volumes_15min_mvt_unfiltered` to match new aggregation method

gabrielwol commented:
> to do:
> 
>     * [x]  insert data during `unacceptable_gaps` into `volumes_15min_mvt_unfiltered` to match new aggregation method
> 
> 
> This was accomplished by simply setting all the nulls to 0 in `volumes_15min_mvt_unfiltered`, since the previous method was to left join unacceptable gaps and set those (padded, zero) volumes to null.

This was incorrect, since the gaps don't have to cover the whole 15 minute period. Wael discovered this error as he was investigating data gaps in Miovision.

Implemented this fix today, (still) running year by year:

```sql
--2019
--UPDATE 150893
--Query returned successfully in 56 secs 272 msec.

--2020
--UPDATE 213973
--Query returned successfully in 1 min 49 secs.

--2021
--UPDATE 136461
--Query returned successfully in 6 min 19 secs.

--2022
--UPDATE 125247
--Query returned successfully in 12 min 22 secs.


WITH v15_gaps AS (
    SELECT datetime_bin_15(datetime_bin) AS datetime_bin, leg, classification_uid, intersection_uid, movement_uid, SUM(volume) AS volume_15min
    FROM miovision_api.volumes
    WHERE
        volume_15min_mvt_uid IS NULL AND datetime_bin >= '2023-01-01' AND datetime_bin < '2024-01-01'
    GROUP BY 1,2,3,4,5
),

updated AS (
    UPDATE miovision_api.volumes_15min_mvt_unfiltered AS v15
    SET volume = v15_gaps.volume_15min
    FROM v15_gaps
    WHERE
        v15.datetime_bin = v15_gaps.datetime_bin
        AND v15.leg = v15_gaps.leg
        AND v15.classification_uid = v15_gaps.classification_uid
        AND v15.intersection_uid = v15_gaps.intersection_uid
        AND v15.movement_uid = v15_gaps.movement_uid
    RETURNING v15.volume_15min_mvt_uid, v15.intersection_uid, v15.datetime_bin, v15.classification_uid, v15.leg, v15.movement_uid
)

UPDATE miovision_api.volumes AS v
SET volume_15min_mvt_uid = v15.volume_15min_mvt_uid
FROM updated AS v15
WHERE v.volume_15min_mvt_uid IS NULL
    AND v.datetime_bin >= '2023-01-01'
    AND v.datetime_bin < '2024-01-01'
    AND v.datetime_bin >= v15.datetime_bin
    AND v.datetime_bin < v15.datetime_bin + interval '15 minutes'
    AND v15.leg = v.leg
    AND v15.classification_uid = v.classification_uid
    AND v15.intersection_uid = v.intersection_uid
    AND v15.movement_uid = v.movement_uid
```
